### PR TITLE
Refactor CacheDatasetOp and add tests

### PR DESCRIPTION
**Pull request by feihugis**
This PR refactors `CacheDatasetOp` and add the unit tests.

cc: @jsimsa

### CODE REVIEW

File: tensorflow/core/kernels/data/cache_dataset_ops.cc [hunk 1]
```diff
@@ -145,578 +275,641 @@ class CacheDatasetOp : public UnaryDatasetOpKernel {
...
+      // iterators to use a common key for `cur_index`. We leverage this
+      // in the corner case when this iterator is restored from an old
+      // checkpoint in `write` mode and the cache has been completely
+      // flushed to disk since then. In that case we simply build a
+      // `FileReaderIterator` and seek to the `cur_index`.
+      switch (mode_) {
+        case Mode::read:
+          iterator_ =
+              absl::make_unique<FileReaderIterator>(FileReaderIterator::Params{
+                  dataset(), strings::StrCat(prefix(), kImpl)});
```
Comment (feihugis) [inline tensorflow/core/kernels/data/cache_dataset_ops.cc]: Note that this iterator prefix is not revised using `name_utils::IteratorPrefix()` as it is different from the prefix patterns we are using (there is no `::` and did not use dataset type as the prefix).

File: tensorflow/core/kernels/data/cache_dataset_ops.cc [hunk 2]
```diff
@@ -145,578 +275,641 @@ class CacheDatasetOp : public UnaryDatasetOpKernel {
...
+          break;
+        case Mode::write:
+          iterator_ =
+              absl::make_unique<FileWriterIterator>(FileWriterIterator::Params{
+                  dataset(), strings::StrCat(prefix(), kImpl)});
```
Comment (feihugis) [inline tensorflow/core/kernels/data/cache_dataset_ops.cc]: Note that this iterator prefix is not revised using `name_utils::IteratorPrefix()` as it is different from the prefix patterns we are using (there is no `::` and did not use dataset type as the prefix).

File: tensorflow/core/kernels/data/cache_dataset_ops.cc [hunk 3]
```diff
@@ -12,128 +12,258 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
 // the following op.
 
-class CacheDatasetOp : public UnaryDatasetOpKernel {
+/* static */ constexpr const char* const CacheDatasetOp::kDatasetType;
+/* static */ constexpr const char* const CacheDatasetOp::kInputDataset;
+/* static */ constexpr const char* const CacheDatasetOp::kFileName;
+/* static */ constexpr const char* const CacheDatasetOp::kOutputTypes;
+/* static */ constexpr const char* const CacheDatasetOp::kOutputShapes;
+
+constexpr char kTensorStrFormat[] = "%%%zuzu_%%%zuzu";
```
Comment (feihugis) [inline tensorflow/core/kernels/data/cache_dataset_ops.cc]: Is this variable name `kTensorStrFormat` OK? Would `kKeyStrFormat` be better

Comment (jsimsa) [reply tensorflow/core/kernels/data/cache_dataset_ops.cc]: `kKeyStrFormat` works for me

**Review by jsimsa — DISMISSED** [review]

**Comment by feihugis** [comment]
@jsimsa The variable name has been updated. Please have a look at the change (https://github.com/tensorflow/tensorflow/pull/30318/commits/e6593c186bed1e740efd4e65463542a1ca3a63d2)! The sanity check failed last time. It seems to be unrelated as the log shows:
```
ERROR: error loading package 'tensorflow/lite/delegates/gpu/metal/kernels': Unable to load package for '//java/com/google/wireless/qa/mobileharness/client/builddefs:mobile_test.bzl': BUILD file not found on package path
INFO: Elapsed time: 7.246s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (53 packages loaded)
FAILED: Build did NOT complete successfully (53 packages loaded)

FAIL: bazel build --nobuild  -- //tensorflow/... -//tensorflow/lite/delegates/gpu/... -//tensorflow/lite/java/demo/app/... -//tensorflow/lite/schema/...
  This is due to invalid BUILD files. See lines above for details.
```

**Review by jsimsa — DISMISSED** [review]

**Comment by feihugis** [comment]
@rthadur The internal checks have been pending for a while. Could you please help trigger the internal checks?

**Comment by feihugis** [comment]
@jsimsa The internal check and Window Bazel check failed. Could you please help check and paste the related log details here?

**Comment by jsimsa** [comment]
The cache dataset ops tests are timing out internally. The following error is repeated in the log a lot of times:

```
third_party/tensorflow/core/kernels/data/cache_dataset_ops_test.cc:172: Failure
Expected equality of these values:
  ::tensorflow::Status::OK()
    Which is: OK
  (iterator->GetNext(iterator_ctx.get(), &next, &end_of_sequence))
    Which is: Permission denied: open failed for cache_data_0.lockfile: Permission denied
```

**Comment by jsimsa** [comment]
When you specify a file path for the test, you should generate it with `env::Default()->LocalTempFilename()`.

**Comment by feihugis** [comment]
@jsimsa Thanks for the suggestion! The `tmp_dir` is added to the specified file path. Could you have a look at the change(https://github.com/tensorflow/tensorflow/pull/30318/commits/436516cacc69f4bbc8c9f66aca742a208af3164e)?

**Review by jsimsa — APPROVED** [review]

**Comment by feihugis** [comment]
Close this PR as the code has been merged. @jsimsa Thanks for your help on this PR!

**Pull request closed without merging**